<a href="https://colab.research.google.com/github/Pvcunha/introducao-ciencia-dados/blob/main/introducao_ciencia_dados.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Airline Passenger Satisfaction

## Lendo dados

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('train.csv')

In [3]:
df.head(5)

Unnamed: 0      id  Gender      Customer Type  Age   Type of Travel  \
0           0   70172    Male     Loyal Customer   13  Personal Travel   
1           1    5047    Male  disloyal Customer   25  Business travel   
2           2  110028  Female     Loyal Customer   26  Business travel   
3           3   24026  Female     Loyal Customer   25  Business travel   
4           4  119299    Male     Loyal Customer   61  Business travel   

      Class  Flight Distance  Inflight wifi service  \
0  Eco Plus              460                      3   
1  Business              235                      3   
2  Business             1142                      2   
3  Business              562                      2   
4  Business              214                      3   

   Departure/Arrival time convenient  ...  Inflight entertainment  \
0                                  4  ...                       5   
1                                  2  ...                       1   
2                                  2  ...                       5   
3                                  5  ...                       2   
4                                  3  ...                       3   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 4                 3                 4                4   
1                 1                 5                 3                1   
2                 4                 3                 4                4   
3                 2                 5                 3                1   
4                 3                 4                 4                3   

   Inflight service  Cleanliness  Departure Delay in Minutes  \
0                 5            5                          25   
1                 4            1                           1   
2                 4            5                           0   
3                 4            2                          11   
4                 3            3                           0   

   Arrival Delay in Minutes             satisfaction  
0                      18.0  neutral or dissatisfied  
1                       6.0  neutral or dissatisfied  
2                       0.0                satisfied  
3                       9.0  neutral or dissatisfied  
4                       0.0                satisfied  

[5 rows x 25 columns]

In [4]:
df.shape

(103904, 25)

## Preprocessing data

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103904 entries, 0 to 103903
Data columns (total 25 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   Unnamed: 0                         103904 non-null  int64  
 1   id                                 103904 non-null  int64  
 2   Gender                             103904 non-null  object 
 3   Customer Type                      103904 non-null  object 
 4   Age                                103904 non-null  int64  
 5   Type of Travel                     103904 non-null  object 
 6   Class                              103904 non-null  object 
 7   Flight Distance                    103904 non-null  int64  
 8   Inflight wifi service              103904 non-null  int64  
 9   Departure/Arrival time convenient  103904 non-null  int64  
 10  Ease of Online booking             103904 non-null  int64  
 11  Gate location                      1039

In [6]:
df.isna().sum()

Unnamed: 0                             0
id                                     0
Gender                                 0
Customer Type                          0
Age                                    0
Type of Travel                         0
Class                                  0
Flight Distance                        0
Inflight wifi service                  0
Departure/Arrival time convenient      0
Ease of Online booking                 0
Gate location                          0
Food and drink                         0
Online boarding                        0
Seat comfort                           0
Inflight entertainment                 0
On-board service                       0
Leg room service                       0
Baggage handling                       0
Checkin service                        0
Inflight service                       0
Cleanliness                            0
Departure Delay in Minutes             0
Arrival Delay in Minutes             310
satisfaction    

Podemos perceber algumas coias a partir dessas informações:
* Arrival Delay in Minutes possui 310 valores faltando.
* As primeiras duas colunas não servem para o nosso contexto.
* Possuímos várias colunas categoricas, tanto numéricas como com Int64.  


In [7]:
df = df.drop(columns=['Unnamed: 0', 'id'])

In [8]:
df.columns

Index(['Gender', 'Customer Type', 'Age', 'Type of Travel', 'Class',
       'Flight Distance', 'Inflight wifi service',
       'Departure/Arrival time convenient', 'Ease of Online booking',
       'Gate location', 'Food and drink', 'Online boarding', 'Seat comfort',
       'Inflight entertainment', 'On-board service', 'Leg room service',
       'Baggage handling', 'Checkin service', 'Inflight service',
       'Cleanliness', 'Departure Delay in Minutes', 'Arrival Delay in Minutes',
       'satisfaction'],
      dtype='object')

Convertendo para tipo categorico todas as colunas exceto, `Age, Flight Distance, Departure Delay in Minutes, Arrival Delay In Minutes, Satisfaction`

In [9]:
except_columns = ['Age', 'Flight Distance', 'Departure Delay in Minutes', 'Arrival Delay in Minutes', 'satisfaction']

In [10]:
# Não existe uma forma de fazer inplace pelo pandas
for col in df.columns:
  if col not in except_columns:
    df[col] = df[col].astype('category')
  

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103904 entries, 0 to 103903
Data columns (total 23 columns):
 #   Column                             Non-Null Count   Dtype   
---  ------                             --------------   -----   
 0   Gender                             103904 non-null  category
 1   Customer Type                      103904 non-null  category
 2   Age                                103904 non-null  int64   
 3   Type of Travel                     103904 non-null  category
 4   Class                              103904 non-null  category
 5   Flight Distance                    103904 non-null  int64   
 6   Inflight wifi service              103904 non-null  category
 7   Departure/Arrival time convenient  103904 non-null  category
 8   Ease of Online booking             103904 non-null  category
 9   Gate location                      103904 non-null  category
 10  Food and drink                     103904 non-null  category
 11  Online boarding           

Describe the information of numerical type

In [12]:
df.describe()

Age  Flight Distance  Departure Delay in Minutes  \
count  103904.000000    103904.000000               103904.000000   
mean       39.379706      1189.448375                   14.815618   
std        15.114964       997.147281                   38.230901   
min         7.000000        31.000000                    0.000000   
25%        27.000000       414.000000                    0.000000   
50%        40.000000       843.000000                    0.000000   
75%        51.000000      1743.000000                   12.000000   
max        85.000000      4983.000000                 1592.000000   

       Arrival Delay in Minutes  
count             103594.000000  
mean                  15.178678  
std                   38.698682  
min                    0.000000  
25%                    0.000000  
50%                    0.000000  
75%                   13.000000  
max                 1584.000000

Descreve informação dos tipos categóricos

In [13]:
df.describe(include=['category'])

Gender   Customer Type   Type of Travel     Class  \
count   103904          103904           103904    103904   
unique       2               2                2         3   
top     Female  Loyal Customer  Business travel  Business   
freq     52727           84923            71655     49665   

        Inflight wifi service  Departure/Arrival time convenient  \
count                  103904                             103904   
unique                      6                                  6   
top                         3                                  4   
freq                    25868                              25546   

        Ease of Online booking  Gate location  Food and drink  \
count                   103904         103904          103904   
unique                       6              6               6   
top                          3              3               4   
freq                     24449          28577           24359   

        Online boarding  Seat comfort  Inflight entertainment  \
count            103904        103904                  103904   
unique                6             6                       6   
top                   4             4                       4   
freq              30762         31765                   29423   

        On-board service  Leg room service  Baggage handling  Checkin service  \
count             103904            103904            103904           103904   
unique                 6                 6                 5                6   
top                    4                 4                 4                4   
freq               30867             28789             37383            29055   

        Inflight service  Cleanliness  
count             103904       103904  
unique                 6            6  
top                    4            4  
freq               37945        27179

### Tratando dados faltantes

Substituimos os dados faltantes pela mediana da coluna para não descartarmos as linhas.

In [14]:
df['Arrival Delay in Minutes'] = df['Arrival Delay in Minutes'].fillna(df['Arrival Delay in Minutes'].median(), axis=0)

In [15]:
df.isna().sum()

Gender                               0
Customer Type                        0
Age                                  0
Type of Travel                       0
Class                                0
Flight Distance                      0
Inflight wifi service                0
Departure/Arrival time convenient    0
Ease of Online booking               0
Gate location                        0
Food and drink                       0
Online boarding                      0
Seat comfort                         0
Inflight entertainment               0
On-board service                     0
Leg room service                     0
Baggage handling                     0
Checkin service                      0
Inflight service                     0
Cleanliness                          0
Departure Delay in Minutes           0
Arrival Delay in Minutes             0
satisfaction                         0
dtype: int64

### Normalização e discretização dos dados

Separando os dados categoricos e numéricos.

In [16]:
numerical_df = df._get_numeric_data().copy()
categorical_df = df.select_dtypes(include=['category']).copy()

#### Normalização de features quantitativas

In [17]:
numerical_df.describe()

Age  Flight Distance  Departure Delay in Minutes  \
count  103904.000000    103904.000000               103904.000000   
mean       39.379706      1189.448375                   14.815618   
std        15.114964       997.147281                   38.230901   
min         7.000000        31.000000                    0.000000   
25%        27.000000       414.000000                    0.000000   
50%        40.000000       843.000000                    0.000000   
75%        51.000000      1743.000000                   12.000000   
max        85.000000      4983.000000                 1592.000000   

       Arrival Delay in Minutes  
count             103904.000000  
mean                  15.133392  
std                   38.649776  
min                    0.000000  
25%                    0.000000  
50%                    0.000000  
75%                   13.000000  
max                 1584.000000

In [18]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler(feature_range=(-1,1)).fit_transform(numerical_df)
scaled_numerical_df = pd.DataFrame(scaler, columns=numerical_df.columns)

In [19]:
scaled_numerical_df.describe()

Age  Flight Distance  Departure Delay in Minutes  \
count  103904.000000    103904.000000               103904.000000   
mean       -0.169751        -0.532129                   -0.981387   
std         0.387563         0.402725                    0.048029   
min        -1.000000        -1.000000                   -1.000000   
25%        -0.487179        -0.845315                   -1.000000   
50%        -0.153846        -0.672052                   -1.000000   
75%         0.128205        -0.308562                   -0.984925   
max         1.000000         1.000000                    1.000000   

       Arrival Delay in Minutes  
count             103904.000000  
mean                  -0.980892  
std                    0.048800  
min                   -1.000000  
25%                   -1.000000  
50%                   -1.000000  
75%                   -0.983586  
max                    1.000000

#### Lidando com dados categóricos

Divideremos em dados features binárias e multivaloradas

In [20]:
binary_cols = [col for col in categorical_df.columns if len(categorical_df[col].unique()) == 2]
nonbinary_cols = [col for col in categorical_df.columns if col not in binary_cols]

In [21]:
binary_df = categorical_df[binary_cols]
binary_df.describe(include=['category'])

Gender   Customer Type   Type of Travel
count   103904          103904           103904
unique       2               2                2
top     Female  Loyal Customer  Business travel
freq     52727           84923            71655

Vamos binarizar as colunas

In [22]:
from sklearn.preprocessing import LabelEncoder

binary_df = binary_df.apply(LabelEncoder().fit_transform)

In [23]:
binary_df.head(5)

Gender  Customer Type  Type of Travel
0       1              0               1
1       1              1               0
2       0              0               0
3       0              0               0
4       1              0               0

Agora vamos fazer o one hot encoding para as outras colunas categoricas multivaloradas

In [24]:
nonbinary_df = categorical_df[nonbinary_cols]
nonbinary_df.describe(include=['category'])

Class  Inflight wifi service  Departure/Arrival time convenient  \
count     103904                 103904                             103904   
unique         3                      6                                  6   
top     Business                      3                                  4   
freq       49665                  25868                              25546   

        Ease of Online booking  Gate location  Food and drink  \
count                   103904         103904          103904   
unique                       6              6               6   
top                          3              3               4   
freq                     24449          28577           24359   

        Online boarding  Seat comfort  Inflight entertainment  \
count            103904        103904                  103904   
unique                6             6                       6   
top                   4             4                       4   
freq              30762         31765                   29423   

        On-board service  Leg room service  Baggage handling  Checkin service  \
count             103904            103904            103904           103904   
unique                 6                 6                 5                6   
top                    4                 4                 4                4   
freq               30867             28789             37383            29055   

        Inflight service  Cleanliness  
count             103904       103904  
unique                 6            6  
top                    4            4  
freq               37945        27179

In [25]:
dummies_nonbinary_df = pd.get_dummies(nonbinary_df)

In [26]:
dummies_nonbinary_df.head(5)

Class_Business  Class_Eco  Class_Eco Plus  Inflight wifi service_0  \
0               0          0               1                        0   
1               1          0               0                        0   
2               1          0               0                        0   
3               1          0               0                        0   
4               1          0               0                        0   

   Inflight wifi service_1  Inflight wifi service_2  Inflight wifi service_3  \
0                        0                        0                        1   
1                        0                        0                        1   
2                        0                        1                        0   
3                        0                        1                        0   
4                        0                        0                        1   

   Inflight wifi service_4  Inflight wifi service_5  \
0                        0                        0   
1                        0                        0   
2                        0                        0   
3                        0                        0   
4                        0                        0   

   Departure/Arrival time convenient_0  ...  Inflight service_2  \
0                                    0  ...                   0   
1                                    0  ...                   0   
2                                    0  ...                   0   
3                                    0  ...                   0   
4                                    0  ...                   0   

   Inflight service_3  Inflight service_4  Inflight service_5  Cleanliness_0  \
0                   0                   0                   1              0   
1                   0                   1                   0              0   
2                   0                   1                   0              0   
3                   0                   1                   0              0   
4                   1                   0                   0              0   

   Cleanliness_1  Cleanliness_2  Cleanliness_3  Cleanliness_4  Cleanliness_5  
0              0              0              0              0              1  
1              1              0              0              0              0  
2              0              0              0              0              1  
3              0              1              0              0              0  
4              0              0              1              0              0  

[5 rows x 86 columns]

### Removendo outliers em dados numéricos

#### Abordagem z-score

O z-score mede a distancia que um valor está longe da média do seu conjunto. para detecção de outliers é utilizado um valor de normalmente +/- 3

In [27]:
from scipy import stats
isoutliers = np.abs(stats.zscore(numerical_df) < 3).all(axis=1)
numerical_df['outliers_zscore'] = np.multiply(isoutliers, 1)

In [28]:
numerical_df.head(5)

Age  Flight Distance  Departure Delay in Minutes  Arrival Delay in Minutes  \
0   13              460                          25                      18.0   
1   25              235                           1                       6.0   
2   26             1142                           0                       0.0   
3   25              562                          11                       9.0   
4   61              214                           0                       0.0   

   outliers_zscore  
0                1  
1                1  
2                1  
3                1  
4                1

In [29]:
numerical_df[numerical_df['outliers_zscore'] == 0]

Age  Flight Distance  Departure Delay in Minutes  \
168      49              578                         162   
178      42             1372                         141   
278      40             1959                         209   
282      26              406                         149   
340      25             1796                         176   
...     ...              ...                         ...   
103770   38             2940                         233   
103812   28              814                         212   
103832   49              639                         225   
103843   42             1076                         171   
103869   18              895                         212   

        Arrival Delay in Minutes  outliers_zscore  
168                        179.0                0  
178                        125.0                0  
278                        191.0                0  
282                        138.0                0  
340                        164.0                0  
...                          ...              ...  
103770                     223.0                0  
103812                     212.0                0  
103832                     222.0                0  
103843                     156.0                0  
103869                     206.0                0  

[2563 rows x 5 columns]

#### Abordagem multivariada (Elliptic Envelope)

In [30]:
from sklearn.covariance import EllipticEnvelope
detector = EllipticEnvelope(contamination=0.01)
detector.fit(numerical_df)

EllipticEnvelope(contamination=0.01)

In [31]:
scores = detector.predict(numerical_df)

In [32]:
scores

array([1, 1, 1, ..., 1, 1, 1])

In [33]:
numerical_df['outliers_multi'] = scores

In [34]:
numerical_df.head()

Age  Flight Distance  Departure Delay in Minutes  Arrival Delay in Minutes  \
0   13              460                          25                      18.0   
1   25              235                           1                       6.0   
2   26             1142                           0                       0.0   
3   25              562                          11                       9.0   
4   61              214                           0                       0.0   

   outliers_zscore  outliers_multi  
0                1               1  
1                1               1  
2                1               1  
3                1               1  
4                1               1

In [36]:
numerical_df[numerical_df['outliers_multi'] == -1]

Age  Flight Distance  Departure Delay in Minutes  \
278      40             1959                         209   
404      25             1258                         199   
582      16             2899                         220   
603      37              107                         232   
665      16             1325                         243   
...     ...              ...                         ...   
103707   57              282                         204   
103770   38             2940                         233   
103812   28              814                         212   
103832   49              639                         225   
103869   18              895                         212   

        Arrival Delay in Minutes  outliers_zscore  outliers_multi  
278                        191.0                0              -1  
404                        185.0                0              -1  
582                        212.0                0              -1  
603                        241.0                0              -1  
665                        226.0                0              -1  
...                          ...              ...             ...  
103707                     202.0                0              -1  
103770                     223.0                0              -1  
103812                     212.0                0              -1  
103832                     222.0                0              -1  
103869                     206.0                0              -1  

[1040 rows x 6 columns]

## Visualização dos dados e analise descritiva

## Teste de hipótese

In [ ]:
from scipy.stats import stats
from scipy.stats import 